src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/ConditionalUpsertResourceHandler.cs

```diff
@@ -0,0 +1,105 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Extensions;
+using Microsoft.Health.Fhir.Core.Features.Conformance;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Messages.Create;
+using Microsoft.Health.Fhir.Core.Messages.Upsert;
+
+namespace Microsoft.Health.Fhir.Core.Features.Resources.Upsert
+{
+    /// <summary>
+    /// Handles Conditional Update logic as defined in the spec https://www.hl7.org/fhir/http.html#cond-update
+    /// </summary>
+    public class ConditionalUpsertResourceHandler : BaseResourceHandler, IRequestHandler<ConditionalUpsertResourceRequest, UpsertResourceResponse>
+    {
+        private readonly ISearchService _searchService;
+        private readonly IMediator _mediator;
+        private readonly bool _featureEnabled;
+
+        public ConditionalUpsertResourceHandler(
+            IFhirDataStore fhirDataStore,
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ISearchService searchService,
+            IMediator mediator,
+            IsEnabled featureEnabled)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory)
+        {
+            EnsureArg.IsNotNull(searchService, nameof(searchService));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(featureEnabled, nameof(featureEnabled));
+
+            _searchService = searchService;
+            _mediator = mediator;
+            _featureEnabled = featureEnabled();
+        }
+
+        public delegate bool IsEnabled();
+
+        public async Task<UpsertResourceResponse> Handle(ConditionalUpsertResourceRequest message, CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNull(message, nameof(message));
+
+            SearchResult results = await _searchService.SearchAsync(message.Resource.InstanceType, message.ConditionalParameters, cancellationToken);
```

Comment:
From what I understand, a server side javascript in Cosmod DB can act within a single transaction (for a single partition). If we used that for the read/update would it avoid the possibility of the contention issue?

Reply:
Stored procedures run on a single partition, the problem here is that the search is cross-partition. In sql we could use a serializable transaction to solve this